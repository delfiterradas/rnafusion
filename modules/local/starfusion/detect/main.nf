process STARFUSION {
    tag "$meta.id"
    label 'process_high'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://community-cr-prod.seqera.io/docker/registry/v2/blobs/sha256/be/bed86145102fdf7e381e1a506a4723676f98b4bbe1db5085d02213cef18525c9/data' :
        'community.wave.seqera.io/library/dfam_hmmer_minimap2_star-fusion:aa3a8e3951498552'}"

    input:
    tuple val(meta), path(reads), path(junction)
    path reference

    output:
    tuple val(meta), path("*.fusion_predictions.tsv")                   , emit: fusions
    tuple val(meta), path("*.abridged.tsv")                             , emit: abridged
    tuple val(meta), path("*.coding_effect.tsv")     , optional: true   , emit: coding_effect
    path "versions.yml"                                                 , emit: versions

    script:
    def prefix = task.ext.prefix ?: "${meta.id}"
    def fastq_arg = reads ? (meta.single_end ? "--left_fq ${reads[0]}" : "--left_fq ${reads[0]} --right_fq ${reads[1]}") : ""
    def junction_arg =  junction ? "-J ${junction}" : ""
    def args = task.ext.args ?: ''
    """
    STAR-Fusion \\
        --genome_lib_dir $reference \\
        $fastq_arg \\
        $junction_arg \\
        --CPU $task.cpus \\
        --examine_coding_effect \\
        --output_dir . \\
        $args

    mv star-fusion.fusion_predictions.tsv ${prefix}.starfusion.fusion_predictions.tsv
    mv star-fusion.fusion_predictions.abridged.tsv ${prefix}.starfusion.abridged.tsv
    mv star-fusion.fusion_predictions.abridged.coding_effect.tsv ${prefix}.starfusion.abridged.coding_effect.tsv

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        STAR-Fusion: \$(STAR-Fusion --version 2>&1 | grep -i 'version' | sed 's/STAR-Fusion version: //')
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    touch ${prefix}.starfusion.fusion_predictions.tsv
    touch ${prefix}.starfusion.abridged.tsv
    touch ${prefix}.starfusion.abridged.coding_effect.tsv
    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        STAR-Fusion: \$(STAR-Fusion --version 2>&1 | grep -i 'version' | sed 's/STAR-Fusion version: //')
    END_VERSIONS
    """
}


