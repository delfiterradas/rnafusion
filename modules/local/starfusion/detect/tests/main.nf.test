nextflow_process {

    name "Test Process STARFUSION"
    script "../main.nf"
    process "STARFUSION"

    setup {
        run("STARFUSION_BUILD") {
            script "../../../starfusion/build/main.nf"
            process {
                """
                input[0] = [
                    [ id:'minigenome fasta' ],
                    file("https://github.com/STAR-Fusion/STAR-Fusion-Tutorial/raw/master/minigenome.fa")
                ]
                input[1] = [
                    [ id:'minigenome gtf' ],
                    file("https://github.com/STAR-Fusion/STAR-Fusion-Tutorial/raw/master/minigenome.gtf")
                ]

                input [2] = file("https://github.com/STAR-Fusion/STAR-Fusion-Tutorial/raw/master/CTAT_HumanFusionLib.mini.dat.gz")
                input [3] = "human"
                """
            }
        }
    }

    test("Should run without failures") {

        when {
            process {
                """
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                    [
                        file("https://github.com/STAR-Fusion/STAR-Fusion-Tutorial/raw/v0.0.1/rnaseq_1.fastq.gz"),
                        file("https://github.com/STAR-Fusion/STAR-Fusion-Tutorial/raw/v0.0.1/rnaseq_2.fastq.gz")
                    ], // reads
                    [] // empty list for junction, as we don't have a pre-computed Chimeric.out.junction file
                ]
                input[1] = STARFUSION_BUILD.out.reference.map { it[1] }
                """
            }
        }

        then {
            assert process.success
            assert snapshot(
                process.out.fusions,
                process.out.abridged,
                process.out.coding_effect,
                process.out.versions
            ).match()
        }

    }

    test("Should create stub files") {

        options "-stub"

        when {
            process {
                """
                input[0] = [
                    [ id:'test_stub', single_end:false ],
                    [
                        file("https://github.com/STAR-Fusion/STAR-Fusion-Tutorial/raw/v0.0.1/rnaseq_1.fastq.gz"),
                        file("https://github.com/STAR-Fusion/STAR-Fusion-Tutorial/raw/v0.0.1/rnaseq_2.fastq.gz")
                    ],
                    [] // empty list for junction
                ]
                input[1] = STARFUSION_BUILD.out.reference.map { it[1] }
                """
            }
        }

        then {
            assert process.success
            assert snapshot(
                process.out
            ).match()
        }

    }

}
