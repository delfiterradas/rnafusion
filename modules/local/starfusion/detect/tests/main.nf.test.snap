{
    "Should create stub files": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_stub",
                            "single_end": false
                        },
                        "test_stub.starfusion.fusion_predictions.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test_stub",
                            "single_end": false
                        },
                        "test_stub.starfusion.abridged.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test_stub",
                            "single_end": false
                        },
                        "test_stub.starfusion.abridged.coding_effect.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "3": [
                    "versions.yml:md5,a8fb6344fdf740dde0941048313fc243"
                ],
                "abridged": [
                    [
                        {
                            "id": "test_stub",
                            "single_end": false
                        },
                        "test_stub.starfusion.abridged.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "coding_effect": [
                    [
                        {
                            "id": "test_stub",
                            "single_end": false
                        },
                        "test_stub.starfusion.abridged.coding_effect.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "fusions": [
                    [
                        {
                            "id": "test_stub",
                            "single_end": false
                        },
                        "test_stub.starfusion.fusion_predictions.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,a8fb6344fdf740dde0941048313fc243"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.3"
        },
        "timestamp": "2024-12-19T18:42:06.988178092"
    },
    "Should run without failures": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test.starfusion.fusion_predictions.tsv:md5,82834fffed743afe07da82bd56d50c99"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test.starfusion.abridged.tsv:md5,d6d20fdd4b5cba21b9c0ebf8e0ea19ff"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test.starfusion.abridged.coding_effect.tsv:md5,95dfce6fdaf3589f23881fe1e855c62b"
                ]
            ],
            [
                "versions.yml:md5,a8fb6344fdf740dde0941048313fc243"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.3"
        },
        "timestamp": "2024-12-19T18:41:49.150362156"
    }
}